### ISSUE TITLE
Parallelise game generation

### ISSUE THREAD

**Issue by JamesCarr1** [JamesCarr1/Connect2_AI#8]
Game generation is currently serialised, i.e:

```
Game 1 is generated:
-> move 1
-> move 2
-> move 3
Game 2 is generated:
-> move 1
-> move 2
-> move 3
-> move 4

etc.
```
This has a number of flaws:

`model.predict()` is called for each move generation. This involves an unsqueeze, a `model.eval()`, a `torch.inference_mode()`, a forward pass (with a batch size of one) and a `model.train()`. This is obviously greatly inefficient.

1. If `model.predict()` was just called once per move, i.e generate ALL board positions for the 1000 games and evaluate them all in one pass, this would significantly speed up performance.
2. `model.predict()` is called for the same position, across multiple MCTSs, many times. The model DOES NOT change in these times, so with `torch.inference_mode()` and `model.eval()`, the result would always be the same. It may be useful to save the results in a hash table.
3. A FULL MCTS is called for the same position many times. The model DOES NOT change in these times, so could also use a dictionary to store these.
4. Considering point 1, it would still be wasteful to call `model.predict()` on every position for the 1000 games. There will be many repeats. For example, at the starting position, ALL 1000 games will initialise with [0, 0, 0, 0]. Rather than compute `model.predict()` on a 1000x4 tensor, just compute it for a 1x4 tensor, and return the result to all MC tree searches. Similarly, in the second position, there are only 4 possible positions. Rather than computing the 1000x4 tensor, we can compute a 4x4 tensor, with no loss of knowledge.

**Comment by JamesCarr1** [comment]
Current idea:

1. Start MCTS. Initialise `n_games` root nodes.
2. Cycle through each root node and find the 'current' position. If it has not yet been passed into the model (i.e not in `computed_positions`), add it to a set. If it has, just pass.
3. Perform a forward pass on the set.
4. Create a dictionary from the set and results, and add it to `computed_positions`.
5. Cycle back through the nodes and perform the MCTS increment, using the output from the dict, updating 'next_states'.
6. Return to 2. for the next `n_sims` moves.
7. Return each MCTS root node.
9. *** Could also add these nodes to another dict, `computed_trees` too. ***
10. Repeat 1-9 until all games have terminated.

**Comment by JamesCarr1** [comment]
Have observed ***significant*** improvement, up to 300x faster:

```
Serially generated and saved 1000 games in 5.3382408618927 seconds
Parallelely generated and saved 1000 games in 0.018381118774414062 seconds
```

**Comment by JamesCarr1** [comment]
For a full training 30 generations:

Notably, the parallel method also produced FAR better results than the serial method ever has (up to 100% accuracy on model value predictions).
Serial:
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
1    0.001    0.001  253.979  253.979 main.py:1(<module>) 
```
Parallel:
```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
1    0.001    0.001   75.871   75.871 main.py:1(<module>)
```

**Comment by JamesCarr1** [comment]
Running for 30 generations on the PC:

Serial:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001  401.431  401.431 main.py:1(<module>)
```

Parallel:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001   31.763   31.763 main.py:1(<module>)
```

Still a 13x improvement!!
